core/src/stored/dir_cmd.cc

```diff
@@ -455,6 +462,45 @@ static bool SetdebugCmd(JobControlRecord* jcr)
   }
 }
 
+static DeviceResource* GetDeviceResource(const std::string& name)
+{
+  return static_cast<DeviceResource*>(
+      my_config->GetResWithName(R_DEVICE, name.c_str()));
+}
+
+static bool SetdeviceCmd(JobControlRecord* jcr)
+{
+  BareosSocket* dir = jcr->dir_bsock;
+
+  Dmsg1(10, "SetdeviceCmd: %s\n", dir->msg);
+
+  std::vector<char> device_name(MAX_SETDEVICE_NAME_LENGTH);
+  std::vector<char> autoselect_value(MAX_SETDEVICE_NAME_LENGTH);
+  int scan = sscanf(dir->msg, setdevice_autoselect, device_name.data(),
+                    autoselect_value.data());
+  if (scan != 2) {
+    dir->fsend(BADcmd, "setdevice", dir->msg);
+    return false;
+  }
+
+  auto res = GetDeviceResource(device_name.data());
+  if (res == nullptr) {
+    std::string err{"Device not found: "};
+    err += device_name.data();
+    dir->fsend(BADcmd, "setdevice", err.c_str());
+    return false;
+  }
+
+  try {
+    BoolString s{autoselect_value.data()};
+    res->autoselect = s.get<bool>();
+    if (res->dev) { res->dev->autoselect = res->autoselect; }
+    return dir->fsend(OKsetdevice, device_name.data(), s.get().c_str());
+  } catch (const std::out_of_range& e) {
+    return dir->fsend(BADcmd, "setdevice", e.what());
+  }
```

Comment:
I wouldn't allow all values for true/false in the protocol.
I'd convert it when the commandline is parsed in the director and then use a common convention when talking to the SD.